### PR TITLE
Compress audit logs

### LINKED ISSUE #63843: log4j2.properties should compress audit logs by default.  maybe all logs.
When audit logs are enabled they can generate a lot of data.  We have been fighting with disk space issues regularly.  It seems like all logs should gzip in the rotation by default since it will create fires that people will respond to in various ways depending on their understanding of log4j or even Elastic.  Of course people need to adjust the logging to suite their needs, but a sane default would be nice.

https://github.com/elastic/elasticsearch/blob/77661af2c5905b16884d1b0d5c7b7c9e86b7bee7/x-pack/plugin/core/src/main/config/log4j2.properties

Our solution was to use a block similar to this.  Feel free to adopt it or share a recommendation.

```
appender.audit_rolling.filePattern = ${sys:es.logs.base_path}${sys:file.separator}${sys:es.logs.cluster_name}_audit-%d{yyyy-MM-dd}-%i.json.gz
appender.audit_rolling.policies.type = Policies
appender.audit_rolling.policies.time.type = TimeBasedTriggeringPolicy
appender.audit_rolling.policies.time.interval = 1
appender.audit_rolling.policies.time.modulate = true
appender.audit_rolling.policies.size.type = SizeBasedTriggeringPolicy
appender.audit_rolling.policies.size.size = 128MB
appender.audit_rolling.strategy.type = DefaultRolloverStrategy
appender.audit_rolling.strategy.fileIndex = nomax
appender.audit_rolling.strategy.action.type = Delete
appender.audit_rolling.strategy.action.basepath = ${sys:es.logs.base_path}
appender.audit_rolling.strategy.action.condition.type = IfFileName
appender.audit_rolling.strategy.action.condition.glob = ${sys:es.logs.cluster_name}_audit*
appender.audit_rolling.strategy.action.condition.nested_condition.type = IfLastModified
appender.audit_rolling.strategy.action.condition.nested_condition.age = 7D
```

Thanks!

### PR DESCRIPTION
**Pull request by pgomulka**
audit logs should be compressed when rolling over due to size based
triggering policy breaching 1GB.
Files are not being deleted.

closes #63843

- Have you signed the [contributor license agreement](https://www.elastic.co/contributor-agreement)?
- Have you followed the [contributor guidelines](https://github.com/elastic/elasticsearch/blob/master/CONTRIBUTING.md)?
- If submitting code, have you built your formula locally prior to submission with `gradle check`?
- If submitting code, is your pull request against master? Unless there is a good reason otherwise, we prefer pull requests against master and will backport as needed.
- If submitting code, have you checked that your submission is for an [OS and architecture that we support](https://www.elastic.co/support/matrix#show_os)?
- If you are submitting this code for a class then read our [policy](https://github.com/elastic/elasticsearch/blob/master/CONTRIBUTING.md#contributing-as-part-of-a-class) for that.

### CODE REVIEW

File: x-pack/plugin/core/src/main/config/log4j2.properties [hunk 1]
```diff
@@ -67,15 +67,19 @@ appender.audit_rolling.layout.pattern = {\
 # "indices" the array of indices that the "action" is acting upon
 # "opaque_id" opaque value conveyed by the "X-Opaque-Id" request header
 # "x_forwarded_for" the addresses from the "X-Forwarded-For" request header, as a verbatim string value (not an array)
-# "transport.profile" name of the transport profile in case this is a "connection_granted" or "connection_denied" event
+# "transport.profile" name of the transport profile iLoggingAuditTrailn case this is a "connection_granted" or "connection_denied" event
```
Comment (tvernum) [inline x-pack/plugin/core/src/main/config/log4j2.properties:70]: ?

Comment (pgomulka) [reply x-pack/plugin/core/src/main/config/log4j2.properties:70]: ah sorry about that. accidental change

**Comment by pgomulka** [comment]
@elasticmachine update branch

**Comment by tvernum** [comment]
I'll leave this review in @albertzaharovits's hands.

File: x-pack/plugin/core/src/main/config/log4j2.properties [hunk 2]
```diff
@@ -71,11 +71,13 @@ appender.audit_rolling.layout.pattern = {\
...
 # "event.category" fixed value "elasticsearch-audit"
 
-appender.audit_rolling.filePattern = ${sys:es.logs.base_path}${sys:file.separator}${sys:es.logs.cluster_name}_audit-%d{yyyy-MM-dd}.json
+appender.audit_rolling.filePattern = ${sys:es.logs.base_path}${sys:file.separator}${sys:es.logs.cluster_name}_audit-%d{yyyy-MM-dd}-%i.json.gz
 appender.audit_rolling.policies.type = Policies
 appender.audit_rolling.policies.time.type = TimeBasedTriggeringPolicy
 appender.audit_rolling.policies.time.interval = 1
 appender.audit_rolling.policies.time.modulate = true
+appender.audit_rolling.policies.size.type = SizeBasedTriggeringPolicy
+appender.audit_rolling.policies.size.size = 1GB
```
Comment (albertzaharovits) [inline x-pack/plugin/core/src/main/config/log4j2.properties:80]: I think we need to explicitly set the `nomax` file index attribute.
```
appender.audit_rolling.strategy.type = DefaultRolloverStrategy
appender.audit_rolling.strategy.fileIndex = nomax
```

Comment (pgomulka) [reply x-pack/plugin/core/src/main/config/log4j2.properties:80]: that's right, good find.

**Comment by albertzaharovits** [comment]
@pgomulka Sorry for the huge delay on this PR.

Because the log file name changes, some folks might consider it breaking,
though only past rolled-over names are changed which I think is key. This change
brings the audit logs more in line with the other logs which is also worth pointing out.
But folks might have automated tools set up for all the logs and we might
get into arguments. To be safe, I would drop a breaking change notice, on `7.x`, on the lines:

```
diff --git a/docs/reference/migration/migrate_7_11.asciidoc b/docs/reference/migration/migrate_7_11.asciidoc
index 20ab1574350..76a9ccbb6f0 100644
--- a/docs/reference/migration/migrate_7_11.asciidoc
+++ b/docs/reference/migration/migrate_7_11.asciidoc
@@ -32,3 +32,26 @@ will be applied to the values of a keyword field; for example, a
 field configured with a lowercase normalizer will return highlighted
 snippets in lower case.
 ====
+
+[discrete]
+[[breaking_711_security_changes]]
+=== Security changes
+
+[[audit-logs-are-rolled-over-and-archived-by-size]]
+.Audit logs are rolled-over and archived by size.
+[%collapsible]
+====
+*Details* +
+In addition to the existing daily rollover, the security audit logs are
+now rolled-over by disk size limit as well. Moreover, the rolled-over logs
+are also gzip compressed.
+
+*Impact* +
+The names of rolled over audit logfiles (but not the name of the current log)
+have changed.
+If you've setup automated tools to consume these files, you must configure them
+to use the new names and to possibly account for gzip archives instead of plaintext.
+The Docker build of Elasticsearch is not affected since it logs on stdout where
+rollover is not performed.
+
+====

```
Similarly for `docs/reference/migration/migrate_8_0.asciidoc` on `master`.

Also for safety I would label the PR as `>breaking` .

Also I would work a bit on the description of the PR, on the same lines with the deprecation notice.

**Comment by albertzaharovits** [comment]
After writing the above I notice you've removed the `7.11` label.
I'm curious for the motivation.

**Comment by pgomulka** [comment]
I removed the 7.11 label as the logs compression would be only enabled in 8.0 exactly for the reason you described https://github.com/elastic/elasticsearch/issues/63843#issuecomment-721566429
This is a breaking change and we should deprecate in 7.x

**Comment by pgomulka** [comment]
@elasticmachine update branch

**Comment by albertzaharovits** [comment]
Alright, then. It's cautious to merge to `master` only and we've been not compressing for a long time and few have complained about it. 

Thank you for solving!

**Review by albertzaharovits — APPROVED** [review]
LGTM

**Pull request merged**